library/core/src/num/dec2flt/parse.rs

```diff
@@ -1,121 +1,246 @@
-/// part, the fractional part, and the exponent in it. Does not handle signs.
-pub fn parse_decimal(s: &str) -> ParseResult<'_> {
-    if s.is_empty() {
-        return Invalid;
+/// Parse the scientific notation component of a float.
+fn parse_scientific(s: &mut AsciiStr<'_>) -> i64 {
+    let start = *s;
+    // SAFETY: the first character is 'e'/'E' and scientific mode is enabled
+    unsafe {
+        s.step();
+    }
+    let mut exp_num = 0_i64;
+    let mut neg_exp = false;
+    if s.first_is2(b'-', b'+') {
+        neg_exp = s.first_is(b'-');
+        // SAFETY: s cannot be empty
+        unsafe {
+            s.step();
+        }
+    }
+    if s.first_isdigit() {
+        s.parse_digits(|digit| {
+            // no overflows here, saturate well before overflow
+            if exp_num < 0x10000 {
+                exp_num = 10 * exp_num + digit as i64;
+            }
+        });
+        if neg_exp { -exp_num } else { exp_num }
+    } else {
+        // ignore 'e' and return back
+        *s = start;
+        0
     }
+}
 
-    let s = s.as_bytes();
-    let (integral, s) = eat_digits(s);
+/// Try to parse a non-special floating point number.
+///
+/// This creates a representation of the float as the
+/// significant digits and the decimal exponent.
+pub fn parse_number(s: &[u8]) -> Option<(Number, usize)> {
+    debug_assert!(!s.is_empty());
 
-    match s.first() {
-        None => Valid(Decimal::new(integral, b"", 0)),
-        Some(&b'e' | &b'E') => {
-            if integral.is_empty() {
-                return Invalid; // No digits before 'e'
-            }
+    let mut s = AsciiStr::new(s);
+    let start = s;
 
-            parse_exp(integral, b"", &s[1..])
-        }
-        Some(&b'.') => {
-            let (fractional, s) = eat_digits(&s[1..]);
-            if integral.is_empty() && fractional.is_empty() {
-                // We require at least a single digit before or after the point.
-                return Invalid;
+    // handle optional +/- sign
+    let mut negative = false;
+    // SAFETY: s cannot be empty due to check in dec2flt.
+    unsafe {
+        if s.first_unchecked() == b'-' {
+            negative = true;
+            if s.step().is_empty() {
+                return None;
             }
+        } else if s.first_unchecked() == b'+' && s.step().is_empty() {
+            return None;
+        }
+    }
+    debug_assert!(!s.is_empty());
 
-            match s.first() {
-                None => Valid(Decimal::new(integral, fractional, 0)),
-                Some(&b'e' | &b'E') => parse_exp(integral, fractional, &s[1..]),
-                _ => Invalid, // Trailing junk after fractional part
-            }
+    // parse initial digits before dot
+    let mut mantissa = 0_u64;
+    let digits_start = s;
+    try_parse_digits(&mut s, &mut mantissa);
+    // SAFETY: s must be larger than digits_start and of the same array
+    let mut n_digits = unsafe { s.offset_from(&digits_start) };
```

Comment:
How do you know this precondition is true? Because `try_parse_digits` can only write to the buffer, not shrink `num_digits`? Please say that in the comment.

Reply:
I've resolved this by making `offset_from` safe.

Reply:
[Fixed](https://github.com/Alexhuszagh/rust/blob/fb1a817fa13f774ce4da27fb9769db2779163030/library/core/src/num/dec2flt/common.rs#L87-L90) by making `offset_from` safe.

```rust
    /// Calculate the offset of slice from another.
    fn offset_from(&self, other: &Self) -> isize {
        other.as_ref().len() as isize - self.as_ref().len() as isize
    }
```